src/ophyd_async/core/detector.py

```diff
@@ -0,0 +1,176 @@
+from .utils import merge_gathered_dicts
+
+
+class DetectorTrigger(Enum):
+    #: Detector generates internal trigger for given rate
+    internal = ()
+    #: Expect a series of arbitrary length trigger signals
+    edge_trigger = ()
+    #: Expect a series of constant width external gate signals
+    constant_gate = ()
+    #: Expect a series of variable width external gate signals
+    variable_gate = ()
+
+
+class DetectorControl(ABC):
+    @abstractmethod
+    def get_deadtime(self, exposure: float) -> float:
+        """For a given exposure, how long should the time between exposures be"""
+
+    @abstractmethod
+    async def arm(
+        self,
+        trigger: DetectorTrigger = DetectorTrigger.internal,
+        num: int = 0,
+        exposure: Optional[float] = None,
+    ) -> AsyncStatus:
+        """Arm the detector and return AsyncStatus.
+
+        Awaiting the return value will wait for num frames to be written.
+        """
+
+    @abstractmethod
+    async def disarm(self):
+        """Disarm the detector"""
+
+
+class DetectorWriter(ABC):
+    @abstractmethod
+    async def open(self, multiplier: int = 1) -> Dict[str, Descriptor]:
+        """Open writer and wait for it to be ready for data.
+
+        Args:
+            multiplier: Each StreamDatum index corresponds to this many
+                written exposures
+
+        Returns:
+            Output for ``describe()``
+        """
+
+    @abstractmethod
+    async def wait_for_index(self, index: int) -> None:
+        """Wait until a specific index is ready to be collected"""
+
+    @abstractmethod
+    async def get_indices_written(self) -> int:
+        """Get the number of indices written"""
+
+    @abstractmethod
+    async def collect_stream_docs(
+        self, indices_written: int
+    ) -> AsyncIterator[Union[StreamResource, StreamDatum]]:
+        """Create Stream docs up to given number written"""
+
+    @abstractmethod
+    async def close(self) -> None:
+        """Close writer and wait for it to be finished"""
+
+
+class StandardDetector(
+    Device,
+    Stageable,
+    Configurable,
+    Readable,
+    Triggerable,
+    WritesExternalAssets,
+):
+    """Detector with useful default behaviour.
+
+    Can be supplied extra devices such that they can also be accessed on the detector.
+    This allows for similar on-the-fly device generation as what is currently done for
+    the ophyd_async.panda.PandA object.
+
+    Must be supplied instances of classes that inherit from DetectorControl and
+    DetectorData, to dictate how the detector will be controlled (i.e. arming and
+    disarming) as well as how the detector data will be written (i.e. opening and
+    closing the writer, and handling data writing indices).
+
+    NOTE: only for step-scans.
+    """
+
+    def __init__(
+        self,
+        control: DetectorControl,
+        data: DetectorWriter,
+        config_sigs: Sequence[SignalR],
+        name: str = "",
+        **plugins: Device,
```

Comment:
I think we decided to ditch `**plugins` and have people inherit from `StandardDetector` instead